File: cli/cmd/agents.go
package cmd

import (
	"fmt"
	"strings"

	"github.com/spf13/cobra"

	"openpa.local/cli/internal/output"
)

func newAgentsCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:     "agents",
		Aliases: []string{"agent"},
		Short:   "Manage A2A and MCP server registrations",
	}
	cmd.AddCommand(
		newAgentsListCmd(),
		newAgentsAddCmd(),
		newAgentsDeleteCmd(),
		newAgentsEnableCmd(true),
		newAgentsEnableCmd(false),
		newAgentsReconnectCmd(),
		newAgentsAuthURLCmd(),
		newAgentsUnlinkCmd(),
		newAgentsConfigCmd(),
	)
	return cmd
}

func newAgentsListCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "list",
		Short: "List registered A2A and MCP tools with auth + enabled status",
		RunE: func(cmd *cobra.Command, _ []string) error {
			if err := requireToken(); err != nil {
				return err
			}
			agents, err := state.Client.ListAgents(cmd.Context())
			if err != nil {
				return err
			}
			if state.Output.JSON {
				return output.PrintJSON(agents)
			}
			t := output.NewTable(state.Output, "TOOL_ID", "TYPE", "ENABLED", "STATUS", "URL")
			for _, a := range agents {
				t.AddRow(
					stringField(a, "tool_id"),
					stringField(a, "agent_type"),
					boolField(a, "enabled", false),
					stringField(a, "status_text"),
					stringField(a, "url"),
				)
			}
			t.Render()
			return nil
		},
	}
}

func newAgentsAddCmd() *cobra.Command {
	var (
		agentType, agentURL, jsonConfig string
		systemPrompt, description       string
		llmProvider, llmModel, effort   string
	)
	cmd := &cobra.Command{
		Use:   "add",
		Short: "Register a new A2A or MCP server",
		Long: `Register a new A2A or MCP server.

  --type a2a --url <url>                       Add an A2A agent
  --type mcp --url <url>                       Add an HTTP/SSE MCP server
  --type mcp --json-config '{<vscode-json>}'   Add an MCP server (stdio or http) from JSON

For MCP servers, additional flags configure the wrapping LLM:
  --system-prompt   Server-specific system prompt
  --description     Display description
  --llm-provider, --llm-model, --reasoning-effort  Per-server LLM overrides`,
		RunE: func(cmd *cobra.Command, _ []string) error {
			if err := requireToken(); err != nil {
				return err
			}
			if agentType == "" {
				return fmt.Errorf("--type is required (a2a or mcp)")
			}
			if agentURL == "" && jsonConfig == "" {
				return fmt.Errorf("either --url or --json-config is required")
			}
			body := map[string]any{"type": agentType}
			if agentURL != "" {
				body["url"] = agentURL
			}
			if jsonConfig != "" {
				body["json_config"] = jsonConfig
			}
			if systemPrompt != "" {
				body["system_prompt"] = systemPrompt
			}
			if description != "" {
				body["description"] = description
			}
			if llmProvider != "" {
				body["llm_provider"] = llmProvider
			}
			if llmModel != "" {
				body["llm_model"] = llmModel
			}
			if effort != "" {
				body["reasoning_effort"] = effort
			}
			agent, err := state.Client.AddAgent(cmd.Context(), body)
			if err != nil {
				return err
			}
			if state.Output.JSON {
				return output.PrintJSON(agent)
			}
			output.PrintKV([][2]string{
				{"tool_id", stringField(agent, "tool_id")},
				{"name", stringField(agent, "name")},
				{"agent_type", stringField(agent, "agent_type")},
				{"url", stringField(agent, "url")},
				{"status_text", stringField(agent, "status_text")},
			})
			return nil
		},
	}
	cmd.Flags().StringVar(&agentType, "type", "", "a2a | mcp")
	cmd.Flags().StringVar(&agentURL, "url", "", "Agent URL")
	cmd.Flags().StringVar(&jsonConfig, "json-config", "",
		"VS Code-style MCP server JSON (for --type mcp)")
	cmd.Flags().StringVar(&systemPrompt, "system-prompt", "", "MCP system prompt")
	cmd.Flags().StringVar(&description, "description", "", "MCP description")
	cmd.Flags().StringVar(&llmProvider, "llm-provider", "", "MCP LLM provider override")
	cmd.Flags().StringVar(&llmModel, "llm-model", "", "MCP LLM model override")
	cmd.Flags().StringVar(&effort, "reasoning-effort", "", "MCP reasoning effort")
	return cmd
}

func newAgentsDeleteCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "delete <tool_id>",
		Short: "Unregister an A2A or MCP tool",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			if err := requireToken(); err != nil {
				return err
			}
			return state.Client.RemoveAgent(cmd.Context(), args[0])
		},
	}
}

func newAgentsEnableCmd(enable bool) *cobra.Command {
	use := "enable"
	short := "Enable an agent for the active profile"
	if !enable {
		use = "disable"
		short = "Disable an agent for the active profile"
	}
	return &cobra.Command{
		Use:   use + " <tool_id>",
		Short: short,
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			if err := requireToken(); err != nil {
				return err
			}
			return state.Client.SetAgentEnabled(cmd.Context(), args[0], enable)
		},
	}
}

func newAgentsReconnectCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "reconnect <tool_id>",
		Short: "Retry a stub agent's connection",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			if err := requireToken(); err != nil {
				return err
			}
			return state.Client.ReconnectAgent(cmd.Context(), args[0])
		},
	}
}

func newAgentsAuthURLCmd() *cobra.Command {
	var returnURL string
	cmd := &cobra.Command{
		Use:   "auth-url <tool_id>",
		Short: "Print the OAuth authorize URL for an agent",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			if err := requireToken(); err != nil {
				return err
			}
			authURL, err := state.Client.GetAgentAuthURL(cmd.Context(), args[0], returnURL)
			if err != nil {
				return err
			}
			if state.Output.JSON {
				return output.PrintJSON(map[string]string{"auth_url": authURL})
			}
			output.Println(authURL)
			return nil
		},
	}
	cmd.Flags().StringVar(&returnURL, "return-url", "",
		"Optional URL to redirect to after the OAuth callback completes")
	return cmd
}

func newAgentsUnlinkCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "unlink <tool_id>",
		Short: "Drop the active profile's OAuth token for an agent",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			if err := requireToken(); err != nil {
				return err
			}
			return state.Client.UnlinkAgent(cmd.Context(), args[0])
		},
	}
}

func newAgentsConfigCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "config",
		Short: "Read or update an MCP / built-in agent's per-profile config",
	}
	cmd.AddCommand(
		&cobra.Command{
			Use:   "get <tool_id>",
			Short: "Show the agent's LLM + meta config",
			Args:  cobra.ExactArgs(1),
			RunE: func(c *cobra.Command, args []string) error {
				if err := requireToken(); err != nil {
					return err
				}
				cfg, err := state.Client.GetAgentConfig(c.Context(), args[0])
				if err != nil {
					return err
				}
				if state.Output.JSON {
					return output.PrintJSON(cfg)
				}
				dumpJSON(cfg)
				return nil
			},
		},
		newAgentsConfigSetCmd(),
	)
	return cmd
}

func newAgentsConfigSetCmd() *cobra.Command {
	var (
		llmProvider, llmModel, effort string
		fullReasoning                 string
		systemPrompt, description     string
	)
	cmd := &cobra.Command{
		Use:   "set <tool_id>",
		Short: "Patch an agent's LLM and meta config (only specified flags change)",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			if err := requireToken(); err != nil {
				return err
			}
			body := map[string]any{}
			if llmProvider != "" {
				body["llm_provider"] = llmProvider
			}
			if llmModel != "" {
				body["llm_model"] = llmModel
			}
			if effort != "" {
				body["reasoning_effort"] = effort
			}
			switch strings.ToLower(fullReasoning) {
			case "":
			case "true":
				body["full_reasoning"] = true
			case "false":
				body["full_reasoning"] = false
			default:
				return fmt.Errorf("--full-reasoning must be 'true' or 'false'")
			}
			if systemPrompt != "" {
				body["system_prompt"] = systemPrompt
			}
			if description != "" {
				body["description"] = description
			}
			if len(body) == 0 {
				return fmt.Errorf("at least one config flag is required")
			}
			return state.Client.UpdateAgentConfig(cmd.Context(), args[0], body)
		},
	}
	cmd.Flags().StringVar(&llmProvider, "llm-provider", "", "LLM provider")
	cmd.Flags().StringVar(&llmModel, "llm-model", "", "LLM model")
	cmd.Flags().StringVar(&effort, "reasoning-effort", "", "low | medium | high")
	cmd.Flags().StringVar(&fullReasoning, "full-reasoning", "", "true | false")
	cmd.Flags().StringVar(&systemPrompt, "system-prompt", "", "System prompt")
	cmd.Flags().StringVar(&description, "description", "", "Description")
	return cmd
}

File: app/cli/commands/agents.py
"""`openpa agents ...` — manage A2A and MCP server registrations.

Mirrors `cli/cmd/agents.go`.
"""

from __future__ import annotations

import json as _json
import sys
from typing import Any, Optional

import typer

from app.cli.commands._helpers import graceful_errors


agents_app = typer.Typer(
    name="agents",
    help="Manage A2A and MCP server registrations.",
    no_args_is_help=True,
)
agents_config_app = typer.Typer(
    name="config",
    help="Read or update an MCP / built-in agent's per-profile config.",
    no_args_is_help=True,
)
agents_app.add_typer(agents_config_app, name="config")


@agents_app.command("list")
@graceful_errors
def agents_list(ctx: typer.Context) -> None:
    """List registered A2A and MCP tools with auth + enabled status."""
    import asyncio

    from app.cli.client._base import Client
    from app.cli.client.agents import list_agents
    from app.cli.config import Config
    from app.cli.output import OutputMode, Table, print_json
    from app.cli.output.formatting import bool_field, string_field

    cfg: Config = ctx.obj["cfg"]
    mode: OutputMode = ctx.obj["mode"]
    cfg.require_token()

    async def _run() -> list[dict[str, Any]]:
        async with Client(cfg) as client:
            return await list_agents(client)

    agents = asyncio.run(_run())

    if mode.json:
        print_json(agents)
        return

    table = Table(mode, "TOOL_ID", "TYPE", "ENABLED", "STATUS", "URL")
    for a in agents:
        table.add_row(
            string_field(a, "tool_id"),
            string_field(a, "agent_type"),
            bool_field(a, "enabled", False),
            string_field(a, "status_text"),
            string_field(a, "url"),
        )
    table.render()


@agents_app.command("add")
@graceful_errors
def agents_add(
    ctx: typer.Context,
    agent_type: Optional[str] = typer.Option(None, "--type", help="a2a | mcp."),
    url: Optional[str] = typer.Option(None, "--url", help="Agent URL."),
    json_config: Optional[str] = typer.Option(
        None, "--json-config",
        help="VS Code-style MCP server JSON (for --type mcp).",
    ),
    system_prompt: Optional[str] = typer.Option(None, "--system-prompt", help="MCP system prompt."),
    description: Optional[str] = typer.Option(None, "--description", help="MCP description."),
    llm_provider: Optional[str] = typer.Option(None, "--llm-provider", help="MCP LLM provider override."),
    llm_model: Optional[str] = typer.Option(None, "--llm-model", help="MCP LLM model override."),
    reasoning_effort: Optional[str] = typer.Option(None, "--reasoning-effort", help="MCP reasoning effort."),
) -> None:
    """Register a new A2A or MCP server."""
    import asyncio

    from app.cli.client._base import Client
    from app.cli.client.agents import add_agent
    from app.cli.config import Config
    from app.cli.output import OutputMode, print_json, print_kv
    from app.cli.output.formatting import string_field

    if not agent_type:
        typer.echo("--type is required (a2a or mcp)", err=True)
        raise typer.Exit(code=1)
    if not url and not json_config:
        typer.echo("either --url or --json-config is required", err=True)
        raise typer.Exit(code=1)

    body: dict[str, Any] = {"type": agent_type}
    if url:
        body["url"] = url
    if json_config:
        body["json_config"] = json_config
    if system_prompt:
        body["system_prompt"] = system_prompt
    if description:
        body["description"] = description
    if llm_provider:
        body["llm_provider"] = llm_provider
    if llm_model:
        body["llm_model"] = llm_model
    if reasoning_effort:
        body["reasoning_effort"] = reasoning_effort

    cfg: Config = ctx.obj["cfg"]
    mode: OutputMode = ctx.obj["mode"]
    cfg.require_token()

    async def _run() -> dict[str, Any]:
        async with Client(cfg) as client:
            return await add_agent(client, body)

    agent = asyncio.run(_run())

    if mode.json:
        print_json(agent)
        return
    print_kv([
        ("tool_id", string_field(agent, "tool_id")),
        ("name", string_field(agent, "name")),
        ("agent_type", string_field(agent, "agent_type")),
        ("url", string_field(agent, "url")),
        ("status_text", string_field(agent, "status_text")),
    ])


@agents_app.command("delete")
@graceful_errors
def agents_delete(
    ctx: typer.Context,
    tool_id: str = typer.Argument(..., help="Tool id."),
) -> None:
    """Unregister an A2A or MCP tool."""
    import asyncio

    from app.cli.client._base import Client
    from app.cli.client.agents import remove_agent
    from app.cli.config import Config

    cfg: Config = ctx.obj["cfg"]
    cfg.require_token()

    async def _run() -> None:
        async with Client(cfg) as client:
            await remove_agent(client, tool_id)

    asyncio.run(_run())


@agents_app.command("enable")
@graceful_errors
def agents_enable(
    ctx: typer.Context,
    tool_id: str = typer.Argument(..., help="Tool id."),
) -> None:
    """Enable an agent for the active profile."""
    _set_agent_enabled(ctx, tool_id, True)


@agents_app.command("disable")
@graceful_errors
def agents_disable(
    ctx: typer.Context,
    tool_id: str = typer.Argument(..., help="Tool id."),
) -> None:
    """Disable an agent for the active profile."""
    _set_agent_enabled(ctx, tool_id, False)


def _set_agent_enabled(ctx: typer.Context, tool_id: str, enabled: bool) -> None:
    import asyncio

    from app.cli.client._base import Client
    from app.cli.client.agents import set_agent_enabled
    from app.cli.config import Config

    cfg: Config = ctx.obj["cfg"]
    cfg.require_token()

    async def _run() -> None:
        async with Client(cfg) as client:
            await set_agent_enabled(client, tool_id, enabled)

    asyncio.run(_run())


@agents_app.command("reconnect")
@graceful_errors
def agents_reconnect(
    ctx: typer.Context,
    tool_id: str = typer.Argument(..., help="Tool id."),
) -> None:
    """Retry a stub agent's connection."""
    import asyncio

    from app.cli.client._base import Client
    from app.cli.client.agents import reconnect_agent
    from app.cli.config import Config

    cfg: Config = ctx.obj["cfg"]
    cfg.require_token()

    async def _run() -> None:
        async with Client(cfg) as client:
            await reconnect_agent(client, tool_id)

    asyncio.run(_run())


@agents_app.command("auth-url")
@graceful_errors
def agents_auth_url(
    ctx: typer.Context,
    tool_id: str = typer.Argument(..., help="Tool id."),
    return_url: Optional[str] = typer.Option(
        None, "--return-url",
        help="Optional URL to redirect to after the OAuth callback completes.",
    ),
) -> None:
    """Print the OAuth authorize URL for an agent."""
    import asyncio

    from app.cli.client._base import Client
    from app.cli.client.agents import get_agent_auth_url
    from app.cli.config import Config
    from app.cli.output import OutputMode, print_json

    cfg: Config = ctx.obj["cfg"]
    mode: OutputMode = ctx.obj["mode"]
    cfg.require_token()

    async def _run() -> str:
        async with Client(cfg) as client:
            return await get_agent_auth_url(client, tool_id, return_url or "")

    auth_url = asyncio.run(_run())

    if mode.json:
        print_json({"auth_url": auth_url})
    else:
        sys.stdout.write(f"{auth_url}\n")


@agents_app.command("unlink")
@graceful_errors
def agents_unlink(
    ctx: typer.Context,
    tool_id: str = typer.Argument(..., help="Tool id."),
) -> None:
    """Drop the active profile's OAuth token for an agent."""
    import asyncio

    from app.cli.client._base import Client
    from app.cli.client.agents import unlink_agent
    from app.cli.config import Config

    cfg: Config = ctx.obj["cfg"]
    cfg.require_token()

    async def _run() -> None:
        async with Client(cfg) as client:
            await unlink_agent(client, tool_id)

    asyncio.run(_run())


# ── config ────────────────────────────────────────────────────────────────

@agents_config_app.command("get")
@graceful_errors
def agents_config_get(
    ctx: typer.Context,
    tool_id: str = typer.Argument(..., help="Tool id."),
) -> None:
    """Show the agent's LLM + meta config."""
    import asyncio

    from app.cli.client._base import Client
    from app.cli.client.agents import get_agent_config
    from app.cli.config import Config
    from app.cli.output import OutputMode, print_json

    cfg: Config = ctx.obj["cfg"]
    mode: OutputMode = ctx.obj["mode"]
    cfg.require_token()

    async def _run() -> dict[str, Any]:
        async with Client(cfg) as client:
            return await get_agent_config(client, tool_id)

    out = asyncio.run(_run())

    if mode.json:
        print_json(out)
    else:
        sys.stdout.write(_json.dumps(out, indent=2, ensure_ascii=False, default=str) + "\n")


@agents_config_app.command("set")
@graceful_errors
def agents_config_set(
    ctx: typer.Context,
    tool_id: str = typer.Argument(..., help="Tool id."),
    llm_provider: Optional[str] = typer.Option(None, "--llm-provider", help="LLM provider."),
    llm_model: Optional[str] = typer.Option(None, "--llm-model", help="LLM model."),
    reasoning_effort: Optional[str] = typer.Option(None, "--reasoning-effort", help="low | medium | high."),
    full_reasoning: Optional[str] = typer.Option(None, "--full-reasoning", help="true | false."),
    system_prompt: Optional[str] = typer.Option(None, "--system-prompt", help="System prompt."),
    description: Optional[str] = typer.Option(None, "--description", help="Description."),
) -> None:
    """Patch an agent's LLM and meta config (only specified flags change)."""
    import asyncio

    from app.cli.client._base import Client
    from app.cli.client.agents import update_agent_config
    from app.cli.config import Config

    body: dict[str, Any] = {}
    if llm_provider:
        body["llm_provider"] = llm_provider
    if llm_model:
        body["llm_model"] = llm_model
    if reasoning_effort:
        body["reasoning_effort"] = reasoning_effort
    if full_reasoning is not None and full_reasoning != "":
        v = full_reasoning.lower()
        if v == "true":
            body["full_reasoning"] = True
        elif v == "false":
            body["full_reasoning"] = False
        else:
            typer.echo("--full-reasoning must be 'true' or 'false'", err=True)
            raise typer.Exit(code=1)
    if system_prompt:
        body["system_prompt"] = system_prompt
    if description:
        body["description"] = description

    if not body:
        typer.echo("at least one config flag is required", err=True)
        raise typer.Exit(code=1)

    cfg: Config = ctx.obj["cfg"]
    cfg.require_token()

    async def _run() -> None:
        async with Client(cfg) as client:
            await update_agent_config(client, tool_id, body)

    asyncio.run(_run())
